### PR TITLE
Add more advance parser to Add command

### PR DESCRIPTION
**Pull request by pablow91**

### CODE REVIEW

**Comment by pablow91** [comment]
It's works by probably needs some refactoring - pls comment :cold_sweat:

File: src/main/kotlin/pl/org/pablo/slack/money/slack/parser/ArgumentParser.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+interface ArgumentParser {
+    fun <R> parse(parser: Parser<R>, command: String): R
+}
+
+@Component
+class ArgumentParserImpl : ArgumentParser {
+    override fun <R> parse(parser: Parser<R>, command: String): R {
+        val commandList = (command.split(" ").toList())
```
Comment (Czocher) [inline src/main/kotlin/pl/org/pablo/slack/money/slack/parser/ArgumentParser.kt]: The parenthesis is not required arround (command ...). Also maybe call it "commands" ;)

File: src/main/kotlin/pl/org/pablo/slack/money/slack/parser/ArgumentParser.kt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        for (c in commandList) {
```
Comment (Czocher) [inline src/main/kotlin/pl/org/pablo/slack/money/slack/parser/ArgumentParser.kt]: Rename "c" -> "command"

File: src/main/kotlin/pl/org/pablo/slack/money/slack/parser/ArgumentParser.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package pl.org.pablo.slack.money.slack.parser
+
+import org.springframework.stereotype.Component
+
+interface Parser<out R> {
```
Comment (Czocher) [inline src/main/kotlin/pl/org/pablo/slack/money/slack/parser/ArgumentParser.kt]: Document please.

File: src/main/kotlin/pl/org/pablo/slack/money/slack/parser/ArgumentParser.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    fun canParse(command: String): Boolean
+    fun canFinalize(): Boolean
+    fun parse(command: String): Boolean
+    fun finalize(): R
+}
+
+interface ArgumentParser {
```
Comment (Czocher) [inline src/main/kotlin/pl/org/pablo/slack/money/slack/parser/ArgumentParser.kt]: Also document.

File: src/main/kotlin/pl/org/pablo/slack/money/slack/parser/AddParameterParser.kt [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            Status.VALUE -> current.users.isNotEmpty()
+            Status.OPTION -> true
+            Status.UNKNOWN -> false
+        }
+    }
+
+    override fun canFinalize(): Boolean = result.isNotEmpty() && current.isEmpty()
+
+    private fun commandType(command: String) = when (command) {
+        in Regex("<@([aA-zZ0-9]+)(\\|[aA-zZ0-9]+)?>") -> Status.USER
```
Comment (Czocher) [inline src/main/kotlin/pl/org/pablo/slack/money/slack/parser/AddParameterParser.kt]: Can you provide those as enum values?

File: src/main/kotlin/pl/org/pablo/slack/money/slack/SlackServiceImpl.kt [hunk 6]
```diff
@@ -11,23 +14,38 @@ class SlackServiceImpl(
...
     override fun add(slackRequest: SlackRequest): String {
-        val text = slackRequest.text.replace(Regex("<@([aA-zZ0-9]+)(\\|[aA-zZ0-9]+)?>"), { "@${it.destructured.component1()}" })
-        val regex = Regex("@([aA-zZ0-9]+) ([0-9]+) ?(.*)?")
-        if (!text.matches(regex)) {
-            throw IllegalArgumentException("Wrong command $text")
-        }
-        val (_, user, value, description) = regex.find(text)!!.groupValues
-        moneyService.addMoney(AddDto(slackRequest.user_id, user, value.toInt(), description))
-        chatService.sendMessage(user, "Added new payment from ${slackRequest.user_name} with value: $value")
+        val arg = argumentParser.parse(addArgumentParser, slackRequest.text)
```
Comment (Czocher) [inline src/main/kotlin/pl/org/pablo/slack/money/slack/SlackServiceImpl.kt]: There's a design patter for this kinds of things: https://sourcemaking.com/design_patterns/interpreter/java/1

Comment (Czocher) [reply src/main/kotlin/pl/org/pablo/slack/money/slack/SlackServiceImpl.kt]: It would solve the repetition with canParse and parse.

Comment (pablow91) [reply src/main/kotlin/pl/org/pablo/slack/money/slack/SlackServiceImpl.kt]: Hmm I'm not sure. `canParse` has one crucial use case that is problematic is when parsing below command - `@U1 10 @U2 20 desc`. Its can be splited into: `P1` `P2` `desc`. The issue is that we know when P1 or P2 ends, but we have no idea what is starting. If we finished `P1` then based on next string value we need to decide whether it's `P2` or `desc`. It's kinda we need to figure out what the `Operator` actually is.

File: src/main/kotlin/pl/org/pablo/slack/money/slack/parser/ArgumentParser.kt (hunk 1 above)
Comment (pablow91) [reply src/main/kotlin/pl/org/pablo/slack/money/slack/parser/ArgumentParser.kt]: Now `command` is whole text received from slack and `element` is single element of that command.

**Pull request merged**